### PR TITLE
Use softplus transform in autoguide

### PR DESCRIPTION
**Pull request by fehiepsi**
This ports @fritzo's PR https://github.com/pyro-ppl/pyro/pull/2767, which can be helpful for the training process, to address #855.

### CODE REVIEW

**Review by fritzo — APPROVED** [review]
Thanks for porting this!

**Comment by fritzo** [comment]
I'm not sure what's causing the errors. Maybe we should preserve the old default
```
scale_constraint = constraints.positive
```
until @vitkl demonstrates improved performance?  I also saw some new failures in https://github.com/pyro-ppl/pyro/pull/2767 and needed to tweak some learning rates.  I'm ok changing also Pyro to preserve the old default until experiments show improvement.

File: numpyro/distributions/constraints.py [hunk 1]
```diff
@@ -369,6 +371,19 @@ def feasible_like(self, prototype):
         return jax.numpy.full_like(prototype, 1 / prototype.shape[-1])
 
 
+class _SoftplusPositive(_GreaterThan):
+    def __init__(self):
+        super().__init__(lower_bound=0.0)
+
+    def feasible_like(self, prototype):
+        return jax.numpy.full(jax.numpy.shape(prototype), np.log(2))
```
Comment (fehiepsi) [inline numpyro/distributions/constraints.py:379]: nit: should we scale softplus transform so that 0 -> 1 (instead of log(2))?

Comment (vitkl) [reply numpyro/distributions/constraints.py:379]: pymc3 initialises to 0, such that `scales = softplus(0)`

https://discourse.pymc.io/t/how-to-choose-initial-values-for-meanfield-sigma-and-transform-rho-sigma/6343/2?u=vitkl

**Comment by vitkl** [comment]
> I also saw some new failures in [pyro-ppl/pyro#2767](https://github.com/pyro-ppl/pyro/pull/2767) and needed to tweak some learning rates. 

Could that be because of initial values `scales = softplus(log(2))` rather than because of learning rates?

**Comment by fehiepsi** [comment]
> Could that be because of initial values scales = softplus(log(2))

The initial value of scale is `0.1` by default. We can change init using `init_scale=...` argument in the constructor of auto guides.

File: numpyro/distributions/constraints.py (hunk 1 above)
Comment (fehiepsi) [reply numpyro/distributions/constraints.py:379]: Interesting! Currently, the behavior in Pyro/NumPyro is to init unconstrained scale parameter to `softplus_inv(0.1)` or `exp_inv(0.1)`. This feasible_like method is only useful to initialize latent variables (e.g. in HMC). I think having 0 -> 1 here is not important...

**Review by fritzo — APPROVED** [review]
LGTM

**Comment by fehiepsi** [comment]
Thanks for reviewing, @fritzo!

**Pull request merged**